File: LeetCode/python/KMP.py
def construct_next(pat):
    n = len(pat)
    N = [-1] * n

    t = -1
    j = 0
    while j < n - 1:
        if t < 0 or pat[j] == pat[t]:
            t += 1
            j += 1
            N[j] = t
        else:
            t = N[t]
    return N


def construct_next_better(pat):
    n = len(pat)
    N = [-1] * n

    t = -1
    j = 0
    while j < n - 1:
        if t < 0 or pat[j] == pat[t]:
            t += 1
            j += 1
            N[j] = t if pat[j] != pat[t] else N[t]
        else:
            t = N[t]
    return N


def kmp(txt, pat):
    m = len(txt)
    n = len(pat)
    N = construct_next_better(pat)
    print(N)
    i = 0
    j = 0
    while i < m and j < n:
        if j < 0 or txt[i] == pat[j]:
            i += 1
            j += 1
        else:
            j = N[j]
    return i - j


if __name__ == "__main__":
    # pat = "ABABC"
    # txt = "CABAABABAC"

    pat = "000010"
    txt = "0001000010"
    print(kmp(txt, pat))

File: LeetCode/cpp/KMP.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

class Solution{
    public:
        vector<int> build_next(string& pat){
            vector<int> dp(pat.size(), -1);
            int t = -1, j = 0;
            while(j < pat.size() - 1){
                if(t < 0 || pat[t] == pat[j]){
                    t++; j++;
                    dp[j] = pat[t] == pat[j] ? dp[t] : t;
                }
                else
                    t = dp[t];
            }
            return dp;
        }

        int kmp(string txt, string pat){
            int m = txt.size(), n = pat.size();
            vector<int> dp = build_next(pat);
            int i = 0, j = 0;
            while(i < m && j < n){
                
                if(j < 0 || pat[j] == txt[i]){
                    i++; j++;
                }
                else
                    j = dp[j];
                if(j == n)
                    return i - j;
            }
            return -1;
        }
};


int main(){
    Solution S;
    string txt = "hello";
    string pat = "ll";
    cout << S.kmp(txt, pat) << endl;
    return 0;
}